Add support for Generals Pilot Logic

MustaphaTR opened the pull request:
Pretty much a copy-paste from DeliversCash stuff. Entering actor gives the levels it has to target. ~~For testcase i gave E1 pilot ability.~~

~~Edit: What Travis isn't happy about is something to do with testcase/mission maps so you can ignore it.~~

MustaphaTR committed:
Add Generals Pilot Logic

abcdefg30 commented on OpenRA.Mods.Common/Activities/DonateExperience.cs:
```diff
@@ -0,0 +1,53 @@
+			if (target.IsDead)
+				return;
+
+			target.TraitOrDefault<GainsExperience>().GiveLevels(level);
```
You should replace the `TraitOrDefault` by just `Trait` here: `target.Trait<GainsExperience>().GiveLevels(level);`
(`TraitOrDefault` would only return `null` if it didn't find anything, and that'd crash then because of the `.GiveLevesls`.)

abcdefg30 commented on OpenRA.Mods.Common/Traits/DeliversExperience.cs:
```diff
@@ -0,0 +1,124 @@
+			if (!order.Queued)
+				self.CancelActivity();
+			
+			var level = self.TraitOrDefault<GainsExperience>().Level;
```
Same case here.

abcdefg30 commented on OpenRA.Mods.Common/Activities/DonateExperience.cs:
```diff
@@ -0,0 +1,53 @@
+		
+		public override Activity Tick(Actor self)
+		{
+			if (target.TraitOrDefault<GainsExperience>().Level == target.TraitOrDefault<GainsExperience>().MaxLevel)
```
Similiar here: `if (target.Trait<GainsExperience>().Level == target.Trait<GainsExperience>().MaxLevel)`

abcdefg30 commented on OpenRA.Mods.Common/Traits/DeliversExperience.cs:
```diff
@@ -0,0 +1,124 @@
+
+		public Order IssueOrder(Actor self, IOrderTargeter order, Target target, bool queued)
+		{
+			if (target.Actor.TraitOrDefault<GainsExperience>().Level == target.Actor.TraitOrDefault<GainsExperience>().MaxLevel || self.TraitOrDefault<GainsExperience>().Level == 0)
```
And here.

abcdefg30 commented on OpenRA.Mods.Common/Traits/DeliversExperience.cs:
```diff
@@ -0,0 +1,124 @@
+			if (target.Type != TargetType.Actor)
+				return;
+			
+			if (target.Actor.TraitOrDefault<GainsExperience>().Level == target.Actor.TraitOrDefault<GainsExperience>().MaxLevel || self.TraitOrDefault<GainsExperience>().Level == 0)
```
And here. ^^

MunWolf commented:
As @abcdefg30 said, especially since you have a requires on the trait so you will never encounter a situation where TraitOrDefault is null.
That makes Trait<> better as it omits some checks that TraitOrDefault<> is forced to do.

MustaphaTR commented:
Updated.

GraionDilach commented on OpenRA.Mods.Common/Traits/DeliversExperience.cs:
```diff
@@ -0,0 +1,127 @@
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Give experience actor has to other actors with the `AcceptsDeliveredExperience` trait.")]
```
This description is unclean. "This actor can grant experience levels equal to it's own current level via entering to other actors with the AcceptsDeliveredExperience trait." perhaps?

GraionDilach commented on OpenRA.Mods.Common/Traits/DeliversExperience.cs:
```diff
@@ -0,0 +1,127 @@
+
+		public Order IssueOrder(Actor self, IOrderTargeter order, Target target, bool queued)
+		{
+			if (target.Actor.Trait<GainsExperience>().Level == target.Actor.Trait<GainsExperience>().MaxLevel || self.Trait<GainsExperience>().Level == 0)
```
I'd consider preventing the targeter to even fire if the actor itself has no experience levels.

MustaphaTR replied:
I couldn't really understand what you mean, preventing a pilot with no level to enter is what `self.Trait<GainsExperience>().Level == 0` part does, which is also checked at L76 and L102.

abcdefg30 replied:
On a different note, we query `self.Trait<GainsExperience>()` in free different places. What about caching it in the constructor?

GraionDilach replied:
```

public IEnumerable<IOrderTargeter> Orders
{
   get
   {
   if (self.Trait<GainsExperience>().Level != 0)
       yield return new DeliversExperienceOrderTargeter();
   }
}

and removing the relevant checks during IssueTarget/ResolveOrder.

MustaphaTR replied:
Sorry for late reply.

It doesn't look like i can use `self` at there like that. Are you sure it is really possible to check that at there?

MunWolf replied:
@MustaphaTR You can pass self through the constructor when you call it in
`public object Create(ActorInitializer init) { return new DeliversExperience(this); }`
and then cache it, I am not saying what @GraionDilach is better since I didn't look but this is how you can do it.

MustaphaTR commented:
Updated.

abcdefg30 commented on OpenRA.Mods.Common/Traits/DeliversExperience.cs:
```diff
@@ -0,0 +1,133 @@
+				if (targetGainsExperience.Level == targetGainsExperience.MaxLevel)
+					return false;
+
+				return targetInfo != null
```
You're already checking `targetInfo == null` above, so this shouldn't be needed.

MustaphaTR commented:
Updated.

abcdefg30 reviewed (dismissed):
You should add a `target != sell` check at the beginning of `CanTargetActor` (and `CanTargetActorFrozenActor`). 👍 after that.

MustaphaTR commented:
Updated and added `target == self` check for `CanTargetActor`. It doesn't look like i can do the same for `CanTargetActorFrozenActor` as appearantly i can't compare a `FrozenActor`. Which i don't think is necessary as your own units can't be frozen (even if they enter fog with no `RevealsShroud:` > `Range: 0c0`) anyway.

MunWolf commented:
@MustaphaTR You can do FrozenActor.Actor to do the check, if you still need to do it.

MustaphaTR commented:
Actually `ValidStances:` tag exists for `AcceptsDeliveredExperience:` so one may set it to enemy (even tho i don't see a reason for that) which can have frozen actors, it's better to check.

MunWolf commented:
A unit which gives the enemy experience but sets it on a timer to explode in a minute ;)
Creative stuff like that I guess.

MustaphaTR commented:
Updated, but for some reason i can't give experience to enemy actors at all even with Enemy on ValidStances:. I'll try to find out why.

MunWolf commented on OpenRA.Mods.Common/Traits/DeliversExperience.cs:
```diff
@@ -0,0 +1,142 @@
+		public class DeliversExperienceOrderTargeter : UnitOrderTargeter
+		{
+			public DeliversExperienceOrderTargeter()
+				: base("DeliverExperience", 5, "enter", false, true) { }
```
You cant target enemy units because you are passing false here to the base.

MustaphaTR commented:
Updated. Hopefully no problems left.

AndriiYukhymchak commented:
Will they be able to get into helicopters/planes on the helipad/airfield?

Mailaender reviewed (dismissed):
Works as promised. :white_check_mark:

abcdefg30 commented:
@MustaphaTR can you remove the testcase now?

MustaphaTR commented:
Removed the testcase.

abcdefg30 commented on OpenRA.Mods.Common/Activities/DonateExperience.cs:
```diff
@@ -0,0 +1,53 @@
+			if (target.IsDead)
+				return;
+
+			target.Trait<GainsExperience>().GiveLevels(level);
```
Sorry, one last thing (now that I think of it). `target` is readonly and won't change therefore, so what about caching its `GainsExperience` trait in the constructor? We could then remove the `.Trait<>` calls from here and from `Tick`.

I made an example for that: https://github.com/abcdefg30/OpenRA/commit/837653862b4b1c7a04e861a20e18d384db1ab8e2
(The example also already caches the trait in `ResolveOrder` from `DeliversExperience` to save 2 more trait lookups.)

You should at least remove the unused `OpenRA.Mods.Common.Effects` and `OpenRA.Traits` usings from this file, I think.

MustaphaTR commented:
Updated.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/66328cc1d2e699039f7e37412e5136dbe81808c7...959b0f3421b18cf74c799926f6653724edb3206c)
